### ISSUE TITLE
Can a tuned GLM-5.3-Flash be served on our stack?

### ISSUE THREAD

**Issue by kreuzhofer** [kreuzhofer/dgx-manager#68]
Part of #65

## Question

**Can a tuned GLM-5.3-Flash actually be served on our stack?** Servability is part of feasibility on this map by decision — an artifact we cannot deploy is not a feasible run.

Two candidate routes, and they have very different costs:

1. **Runtime LoRA in vLLM — no merge at all.** Does vLLM's `--enable-lora` path work against an **NVFP4 `glm5_next` base**? Specifically: does vLLM support LoRA on a quantized base at all for this quant scheme; does it support LoRA on **MoE expert weights** (historically a gap — LoRA has often been restricted to attention on MoE models); does it work with our pinned `--moe-backend marlin`; and does it survive alongside `--speculative-config mtp`? Also what it costs in KV pool and throughput. This route would make the whole merge/re-quantize problem vanish, so establish it first.
2. **Merge, then re-quantize to NVFP4.** If merge is required: what does merging an adapter into a ~320B MoE cost in RAM/disk/time, and what does re-quantizing it to NVFP4 take? #48 established the equivalent path for Qwen3.8-27B on the 5090 (NVFP4 on MLP, FP8 on attention, BF16 on vision + MTP; `model_free_ptq` to preserve MTP, because anything routing through `transformers` silently drops it via `_keys_to_ignore_on_load_unexpected`). Does `llmcompressor` support `glm5_next`? Can `LibertAIDAI`'s NVFP4 recipe be reproduced, or is it undocumented? Is there hardware anywhere we have that can hold a 640 GB BF16 model to merge into?
3. **If the base was quantized during training** (QLoRA over NVFP4), is "merge" even well-defined, and what is the accepted practice?
4. **MTP survival.** The MTP head is worth **1.57–1.82× decode** on this model. Does either route preserve it?

**Answer shape.** A verdict on route 1 (works / does not / works with caveats), and if it fails, a costed description of route 2 including whether we own hardware that can execute it. Name the vLLM version each claim is true for.

**Sources.** vLLM docs and source for LoRA + quantization + MoE interaction; `llmcompressor` model support; memory `glm53-flash-is-a-different-model`; #48's write-up.

**Comment by kreuzhofer** [comment]
**Findings:** [`docs/research/glm53-flash-serving-tuned.md`](https://github.com/kreuzhofer/dgx-manager/blob/research/glm53-flash-serving-tuned/docs/research/glm53-flash-serving-tuned.md) on branch `research/glm53-flash-serving-tuned`.

## Verdict: route 1 does not work — but for none of the four reasons this ticket expected. It is ~30 lines in one upstream file.

**Every kernel-level worry in Q1 is already solved, and has been since vLLM v0.25.0 (2026-07-11)** — verified by fetching the files at `v0.25.0`, `v0.26.0`, `v0.27.0`, `v0.27.1`, `v0.28.0` and `main` @ `85c1365`:

- **LoRA on a quantized base:** scheme-agnostic by construction. `base_linear.py:204` calls the base layer's own quantized GEMM and adds an unquantised delta; there is no per-scheme allowlist and no quantization check in `LoRAConfig.verify_with_model_config`. Compressed-tensors is explicitly anticipated — the `weight` property has a `weight_packed` branch, which is exactly NVFP4's storage.
- **LoRA on MoE experts:** the "attention only" limitation is gone. `FusedMoEWithLoRA` / `FusedMoE3DWithLoRA` exist with per-expert Triton kernels.
- **`--moe-backend marlin`:** `class MarlinExperts(LoRAExpertsMixin, MarlinExpertsBase)`, and `NvFp4MoeBackend.MARLIN` resolves to exactly that class. **The flag we pin as mandatory is the LoRA-capable backend.** Bonus: the oracle *filters out* kernels that can't do LoRA, so the FLASHINFER_CUTLASS backend that silently corrupts our output becomes unreachable under `--enable-lora`.
- **MTP:** the only spec-decode/LoRA incompatibility in the whole tree is `enable_adaptive_verification`, which we don't use.

**The blocker is the model file:**

```python
# vllm/models/glm5next/nvidia/model.py:866   (PR #53906, unchanged since its first commit)
class Glm5NextForCausalLM(nn.Module, HasInnerState, SupportsPP, MixtureOfExperts, IsHybrid):
```

No `SupportsLoRA`, no `packed_modules_mapping`, no `lora_skip_prefixes`. The peer model `DeepseekV4ForCausalLM` — also MLA + huge MoE + FP4 experts + MTP — declares all three and is served with LoRA today. **GLM-5.3-Flash's vLLM support is an unmerged five-day-old PR ([#53906](https://github.com/vllm-project/vllm/pull/53906)); `glm5_next` is in no vLLM release.**

**And it fails silently, which is the part to carry forward.** `Glm5NextForConditionalGeneration` inherits from `Glm4vForConditionalGeneration`, which **is** `SupportsLoRA` — so `--enable-lora` passes the interface gate, the server starts, and it runs on GLM-4V's `packed_modules_mapping` (`qkv_proj: [q_proj,k_proj,v_proj]`), naming modules a NoPE-MLA model doesn't have. Tracing the loader: `packed_modules_mapping.get(parts, [])` returns `[]` for the two *fused* projections — `fused_qkv_a_proj` (11 sparse-MLA layers) and `in_proj_qkvbfg_a` (34 KDA layers) — and every candidate wrapper needs a non-empty packed list, so **LoRA is silently absent from all 45 layers' input projections**. (Also: `q_conv1d`/`k_conv1d`/`v_conv1d` are declared `ColumnParallelLinear`, so `target_modules="all-linear"` would try to adapt convolution weights.)

### Q5 — cost. This is where route 1 stops being free even after the code lands

LoRA buffers are allocated at model-load and profiled, so they come out of the KV pool. Derived from vLLM's tensor shapes × the config (288 experts × 42 MoE layers, `fully_sharded_loras` off so w13 `lora_A` isn't sharded):

| rank | all-expert MoE LoRA, per rank @ TP2 | attention only | ratio |
|---:|---:|---:|---:|
| 8 | **2.77 GiB** | 7.4 MiB | 384× |
| 16 | **5.54 GiB** | 14.8 MiB | 384× |

Against our *measured* pool at gmu 0.87: **6.85 GiB without MTP, 3.74 GiB with**. So a single rank-16 all-expert adapter is **1.5× the entire KV pool we run with MTP on**. `max_lora_rank` is a `Literal[1,8,16,32,…]` — there is no rank-4 escape. Ways out: adapt a subset of experts (384× cheaper), `--enable-moe-shared-loras` (~5×, but the adapter must be *trained* in that layout), or give back MTP. That is an independent argument for the map's **expert-subset tuning** option. Throughput cost: mechanism described, no number — nothing was run.

### Q4 — MTP survives route 1, but the drafter is not adapted

`vllm/v1/spec_decode/` contains **zero** LoRA references; the MTP head is a separately registered model (`Glm5NextMTPModel` → `Glm5NextMTP`, no `SupportsLoRA`). Upstream's convention is explicit — `deepseek_v4`: `lora_skip_prefixes = ["mtp."]  # The MTP draft head is not LoRA-adapted`. So draft comes from the unadapted base, verify from base+adapter: **correctness is safe** (rejection sampling), but acceptance falls by an unmeasured amount, eating into our 1.57–1.82×. Trap for whoever writes the patch: copying `["mtp."]` verbatim skips *nothing* — GLM names the head positionally as layer 45.

## Q2/Q3 — route 2, costed, and we can just barely execute it

**First, a correction to the map:** `zai-org/GLM-5.3-Flash` is **FP8** (328.34 GB, tagged `fp8`), not BF16. The BF16 weights are a separate published repo, `zai-org/GLM-5.3-Flash-BF16` — 642.65 GB / 598.52 GiB, 120 shards. That's the merge target, and you don't have to produce it.

| step | cost | risk |
|---|---|---|
| Download BF16 base | 642.65 GB | none |
| Merge adapter **in place**, shard by shard | overwrite | **no first-party streaming merger.** PEFT's `onload_layer` offload-merge is CI-tested (`test_offload_merge`) — on a tiny GPT-2, never at 320B. The only true streaming merge is third-party (`unsloth-zoo`), untested on `glm5_next`. `mergekit` is the wrong tool (it just calls PEFT) |
| `model_free_ptq` → NVFP4, deleting BF16 shards as they convert | writes ~190–198 GB | needs a **dev-mode source install**: released `llmcompressor` 0.13.0 pins `transformers<=5.14.1`, and `glm5_next` first exists in **5.16.1** (verified by tag probing) |

**Peak disk ≈ 660 GB. Memory is never the constraint** — `_no_split_modules` makes the largest indivisible unit one decoder layer, **13.81 GiB at BF16**; `model_free_ptq` wants ~16 GB/worker.

**Hardware: exactly one machine.** Measured today — `dgx-spark-04` local NVMe **714 G free**; `agenthost` 504 G (and only 30 GB RAM); `/mnt/tank` **340 G free of 3.6 T**; `dgx-spark-01` 80 G. So yes, we can do it, on spark-04, at a 92% fill on its root filesystem. Buying a disk is the honest answer if this becomes real.

**MTP in route 2 is a triple silent failure**, worse than #48's:
1. `transformers` drops it on load by **hardcoded index** — `_keys_to_ignore_on_load_unexpected = [r"layers\.45\.", …]`;
2. `llm-compressor`'s copy-back gates on `num_mtp_layers`/`mtp_num_hidden_layers`; GLM's field is `num_nextn_predict_layers`, so it never fires;
3. even if it did, `compressed-tensors`' default `mtp_prefix="mtp"` matches **zero** GLM tensors.

llm-compressor PR [#3118](https://github.com/vllm-project/llm-compressor/pull/3118) (opened 2026-08-31, open) fixes all three. `model_free_ptq` remains the #48-style escape — it never constructs a transformers model, so layer 45 is simply visible — and it *does* support NVFP4 weight-only, which is exactly the split LibertAI ships.

**Is LibertAI's recipe reproducible? No.** It's NVIDIA ModelOpt with no published recipe or command; you can only reconstruct the split from `config.json` (one 4-bit `Linear` group, `input_activations: null`, 43-entry ignore ⇒ only `mlp.experts.*` at NVFP4). **`RedHatAI/GLM-5.3-Flash-NVFP4` does publish `recipe.yaml`** (compressed-tensors) — but it doesn't reproduce its own checkpoint: the recipe has one config group, the shipped config has two, the second putting layer 45 at FP8-block. Of ~100 GLM-5.3-Flash quants on the Hub, two publish a recipe.

**Q3 — is "merge" defined over a quantized base?** Only for bitsandbytes 4-bit, and there it's dequantize→add→requantize with an explicit warning. GPTQ refuses, AWQ has no `merge` at all, HQQ dequantizes. **PEFT has zero NVFP4/compressed-tensors integration** (`grep -rni nvfp4` → no hits) and no block-scaled-FP8 LoRA layer, so you can't merge into the FP8 release either. Accepted practice per PEFT's own docs and unsloth's code: merge into the full-precision sibling. Since `-BF16` is published, the question resolves without needing an answer.

## ⚠ One finding outside this ticket that lands on production

vLLM issue [#54150](https://github.com/vllm-project/vllm/issues/54150) (open): ModelOpt NVFP4 conversions of GLM-5.3-Flash emit **invalid UTF-8 byte tokens on SM120** — 86 / 62 / 94 U+FFFD over six runs, including **the exact `LibertAIDAI/GLM-5.3-Flash-NVFP4` we serve** — versus **0** for RedHat's compressed-tensors build. At `temperature=0`, at zero context, MTP on or off, **with `moe_backend: marlin`**. A comment identifies the cause in vLLM: ModelOpt carries separate global scales for w1/w3 and `modelopt.py` picks one, mis-scaling the other *"up to 10× on GLM-5.3-Flash NVFP4"*. vLLM logs it as a `warning_once`: `"w1_weight_scale_2 must match w3_weight_scale_2. Accuracy may be affected."`

**Unverified on our fleet** (`-02`/`-03` are powered down), but cheap: grep the next GLM-5.3-Flash boot log. If it fires, re-quantise toward **compressed-tensors**, not toward LibertAI's ModelOpt output — and our GPQA numbers are English-only and would not have surfaced this.

## Recommendation

**Do the upstream PR, not route 2.** Adding `SupportsLoRA`, a real `packed_modules_mapping` for the MLA/KDA fusions, and a `lora_skip_prefixes` for the positionally-named MTP head to `vllm/models/glm5next/nvidia/model.py` is ~30 lines against a tree you can name, it deletes route 2 entirely, and every hard part is already built. **Do not attempt it as a ninth overlay layer** — the day-0 image is a local dev build (`ai.vllm.build.commit: "unknown"`, `local/vllm-openai:dev`) created nine hours *before* PR #53906's first commit, so it cannot be diffed against any known tree.

**Not proven, and cheap to find out:** nothing here was executed. One deploy of `--enable-lora` on `-02`/`-03` settles both the "does it start / what is silently skipped" question and the UTF-8 corruption question from the same boot log.
